[RuntimeClass scheduling] native scheduler support, ready to implement

tallclair opened the pull request:
Updated the RuntimeClass scheduling KEP to use native scheduler support rather than a mixin approach.

This approach uses a cleaner API (doesn't need to mixin a bunch of rules into the pod's node affinity), it let's us use the more expressive `NodeSelector` API, and provides a nice user experience in the error case.

I am opening this PR to continue the discussion @bsalamat started [here](https://github.com/kubernetes/enhancements/pull/896#discussion_r267141010)

/sig node scheduling
/assign @bsalamat @derekwaynecarr @dchen1107 
/cc @yastij 
/milestone v1.15
/priority important-longterm

yastij commented on keps/sig-node/runtime-class-scheduling.md:
```diff
@@ -190,43 +181,26 @@ inculde:
-scheduling is to merge the topology constraints from the RuntimeClass into the
-PodSpec. Eventually, the controller's responsibilities may grow, such as to
-merge in [pod overhead][] or validate feature compatibility.
+A new scheduler predicate will manage the RuntimeClass topology. It will lookup
```
Maybe we should discuss predicate composition vs adding support of runtimeClass into other existing predicate

tallclair replied:
If we add it to existing predicates, we don't get a useful error message when the pod can't be scheduled, right?

yastij replied:
Yes

yastij replied:
@tallclair - if we merge tolerations into podSpec how can we inform user correctly (UX) when not finding a node that fits ?

tallclair replied:
I don't think that changes anything, as the tolerations would be resolved by the same `PodToleratesNodeTaints` predicate? But I might be misunderstanding how the debug messages are created.

yastij replied:
Do you mean that we’d still check the runtimeClass when computing `podToleratesNodeTaints` ?

yastij replied:
I think we can move forward, and discuss this while implementing

yastij commented on keps/sig-node/runtime-class-scheduling.md:
```diff
@@ -190,43 +181,26 @@ inculde:
-#### Mix-in
-
-The RuntimeClass topology is merged with the pod's NodeSelector & Tolerations.
+The `PodToleratesNodeTaints` predicate will also be extended to be
```
Can't we just call from the `RuntimeClassPredicate` the `PodToleratesNodeTaints` with the right parameters ? composing seems easier than modifying each predicate to include runtimeClass awareness

tallclair replied:
I assume predicates are AND'd together, in which case the PodToleratesNodeTaints predicate would also be called without the runtimeclass tolerations, and in that instance fail? I'm not very familiar with the scheduler code though, so I could be totally wrong.

yastij replied:
This raises a broader Issue at predicates level, when a predicate is basically a composition of other:

- do we add domain awareness to the existing predicates ?
- do we compose these predicates into domain predicate ?

This might over complicate our predicates definition, and create performance overhead in some cases.

tallclair commented:
Summarizing feedback from 2019-03-26 sig-node:

- Add a section about node label restrictions, and how those should play into the RuntimeClass topology (security considerations)
- Additional tradeoff on native scheduling vs. mixin: with native scheduling, it's harder to apply scheduling policy. This seems like more of a concern with tolerations than with the node selector, so I think we should actually take a hybrid approach, where tolerations are mixed in but the node selector is applied at scheduling time.

/assign @egernst

yastij commented:
@tallclair -  I think this has 2 main benefits

- This tradeoff aligns with the scheduling policy, especially that enforcement of affinities will be done at scheduling time.

- This will avoid the issue of predicate composition (If we still want to go down this path, I think we need a KEP for predicate composition)

my only concern is on the UX side as @bsalamat stated

bsalamat commented on keps/sig-node/runtime-class-scheduling.md:
```diff
@@ -138,13 +138,12 @@ The RuntimeClass definition is augmented with an optional `Topology` struct:
-    NodeSelectorTerm []corev1.NodeSelectorRequirement
+    NodeSelector corev1.NodeSelector
 
     // tolerations adds tolerations to pods running with this RuntimeClass.
```
I am a bit unsure about having tolerations as a part of "topology" in RuntimeClass. This is useful if one doesn't want to provide RuntimeClass on all their pods. If they provide RuntimeClass for all their pods they can just use NodeSelector. If they insist on tainting some nodes, they can use regular Toleration (which is not a part of RuntimeClass). So, I guess my question is "what is the major benefit of having tolerations" here?

tallclair replied:
I think this is covered under the Tolerations discussion on line 170:

> **Tolerations**
> 
> While NodeSelectors and labels are used for steering pods _towards_ nodes,
> [taints and tolerations][taint-and-toleration] are used for steering pods _away_
> from nodes. If every pod had a RuntimeClass and every RuntimeClass had a strict
> NodeSelector, then RuntimeClasses could use non-overlapping selectors in place
> of taints & tolerations. However the same could be said of regular pod
> selectors, yet taints & tolerations are still a useful addition. Examples of
> [use cases](#user-stories) for including tolerations in RuntimeClass topology
> inculde:
> 
> - Tainting Windows nodes with `windows.microsoft.com` to keep default linux pods
>   away from the nodes. Windows RuntimeClasses would then need a corresponding
>   toleration.
> - Tainting "sandbox" nodes with `sandboxed.kubernetes.io` to keep services
>   providing privileged node features away from sandboxed workloads. Sandboxed
>   RuntimeClasses would need a toleration to enable them to be run on those
>   nodes.

I agree that it's not strictly necessary, but I think it will be useful in practice. I guess both of the use cases listed could be accomplished with a `default` RuntimeClass that looked like this:

```
kind: RuntimeClass
metadata:
  name: default
handler: runc
nodeSelector:
  nodeSelectorTerms:
    matchExpressions:
      - key: "windows.microsoft.com"
        operator: DoesNotExist
      - key: "sandboxed.kubernetes.io"
        operator: DoesNotExist
```

yastij replied:
Does this assume that the toleration key is also set as label ? I think that this isn't always the case.

tallclair replied:
This implies using labels rather than taints.

yastij replied:
it is more staightforward for users to implement dedicated nodes using taints and tolerations rather than nodeSelectors. 
The benefit of having tolerations on the runtimeClass is freeing users that want to consume a specific runtimeClass available on a dedicated node to think about which toleration they should set or not.

bsalamat replied:
Your points are valid. Thanks for clarifying.

tallclair commented:
Integrated feedback from sig-node. In particular, the RuntimeController was added back, and tolerations are back to being a mixin.

bsalamat commented on keps/sig-node/runtime-class-scheduling.md:
````diff
@@ -203,6 +202,36 @@ will be a clear explanation of why. For example:
 0/10 Nodes are available: 5 nodes do not have enough memory, 5 nodes don't match RuntimeClass
 ```
 
+### RuntimeController
````
nit: I would call this `Runtime Admission Controller` to clarify that this is not a controller with a control loop that runs all the time and monitors things in the cluster.

yastij replied:
I agree

tallclair replied:
Looks like the convention of the admission plugins is to just omit the controller piece altogether - so this would just be "Runtime". I'll clarify the name in the KEP though.

yastij replied:
or rename to `RuntimeTopology` ?

tallclair replied:
The plan is to reuse the controller for other RuntimeClass-related admission controls. In particular PodOverhead. So I don't think it makes sense to include topology in the name.

derekwaynecarr replied:
agree with a single admission plugin for a common feature.

egernst replied:
@tallclair - have a suggested controller name in mind?  ContainerRuntime? Runtime is pretty overloaded...

tallclair replied:
I went with `RuntimeClass` for the controller name. Pretty self-explanatory.

Huang-Wei commented on keps/sig-node/runtime-class-scheduling.md:
```diff
@@ -138,13 +138,12 @@ The RuntimeClass definition is augmented with an optional `Topology` struct:
+    // node affinity or node selector requirements.
     // +optional
-    NodeSelectorTerm []corev1.NodeSelectorRequirement
+    NodeSelector corev1.NodeSelector
```
Is this change due to the preference of adopting a new scheduler predicate, so that we don't need to mix-in this info?

tallclair replied:
Yes. If we don't need to worry about logically merging this with the existing pod's node affinity, then I prefer the more expressive `NodeSelector` over a `NodeSelectorTerm`. I could be convinced otherwise though.

yastij replied:
not really opposed this.

yastij commented:
/lgtm
/hold

just a rename or clarification on the name of the admission controller is missing.

@bsalamat - any thoughts on merging toleration into podSpec ? This would help enforcing `schedulingPolicy`

bsalamat commented:
> @bsalamat - any thoughts on merging toleration into podSpec ? This would help enforcing schedulingPolicy

Are you saying that instead of having "toleration" in the RuntimeClass, we use the same toleration of Podspec? If so, we discussed it in this KEP and decided that the existence of toleration in the runtime class is useful.

tallclair commented:
@bsalamat I think @yastij was asking about your thoughts on merging the tolerations FROM the RuntimeClass into the PodSpec during admission (what this KEP currently proposes), as opposed to just merging the tolerations at scheduling time (what it originally proposed).

yastij commented:
@bsalamat - I agree 

> @tallclair - if we merge tolerations into podSpec how can we inform user correctly (UX) when not finding a node that fits ?

I think we can keep them separate for now. If we see that we need to add tolerations to the podSpec we can do it afterwards.

derekwaynecarr commented:
@tallclair reached out and asked that i review.  i am fine with what is presented.

yastij commented:
@tallclair @derekwaynecarr  - I think we can drop the renaming, you are points are valid.

On the scope of the admission controller I really think that we shouldn’t merge tolerations, otherwise we’ll loose UX benefits, as we won’t know from where tolerations came; original podspec or runtimeClass, assuming we don’t check the runtimeClass at the predicate level.

Aside from this, I think we can merge this as is.

egernst commented on keps/sig-node/runtime-class-scheduling.md:
```diff
@@ -190,42 +186,51 @@ inculde:
+   is narrowing the scheduling domain rather than expanding it.
+
 [merge-tolerations]: https://github.com/kubernetes/kubernetes/blob/58021216b16ae6d5f24fb1c32ab541b2e79a365e/pkg/util/tolerations/tolerations.go#L62
+[TaintBasedEvictions]: https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#taint-based-evictions
```
while you are here, typo on line 233/238:
s/depnedent/dependent

egernst commented:
@yastij @tallclair - 
`On the scope of the admission controller I really think that we shouldn’t merge tolerations, otherwise we’ll loose UX benefits, as we won’t know from where tolerations came; original podspec or runtimeClass, assuming we don’t check the runtimeClass at the predicate level.` 

What happens if the runtimeClass is removed and the pod is still running?  The toleration won't be tied to the pod any longer?

yastij commented:
I agree that we need to think about taintBasedEviction. This would likely make `taint_manager` aware of the runtimeClass. More generally this might require making any decision making around taint and toleration aware of runtimeClass.

This would generate a significant refactoring of `taint_manager`. If we merge the tolerations into the podSpec, we should check the runtimeClass at the predicate level to keep the UX we want. this has the benefit of not leaking the runtimeClass to the `taint_manager` or any other decision making around tolerations.

tallclair commented:
I spoke with @bsalamat offline, and we agreed that it makes sense to merge the tolerations into the pod spec, but the `RuntimeClass.Topology.NodeSelector` will still be handled in a new scheduler predicate. There isn't a UX benefit to handling the RuntimeClass tolerations in the scheduler, because they still need to be handled in the same PodHandlesNodeTaints predicate, meaning the error message is still the same.

If the user experience is the same, merging the tolerations is much simpler (for the reasons outlined about, around handling RC deletion, etc.), and also makes it easier to apply toleration restrictions through policies (e.g. SchedulingPolicy, gatekeeper).

tallclair commented:
@yastij WDYT? Does that alleviate your concerns?

yastij commented:
@tallclair - sgtm

yastij commented:
/hold cancel

bsalamat commented:
@tallclair
@egernst has a good point. What will happen if a RuntimeClass is deleted after pod creation? I would like to clarify the behavior of the scheduler for pods that are created but not scheduled yet.  I am not concerned about already running pods.
When the scheduler sees a pod with a non-existing RuntimeClass, what should it do? I think it is reasonable to assume that no RuntimeClass is specified and always pass the predicate. Another obvious option is to always fail the RuntimeClass predicate and never schedule the pod. Anyway, I think this should be specified in the KEP.

tallclair commented:
Hmm, good point. I wasn't thinking about the case that the RuntimeClass was missing before the pod is scheduled...

> Another obvious option is to always fail the RuntimeClass predicate and never schedule the pod.

I think this makes the most sense, otherwise we're just delaying the failure until it lands on a node. If the predicate fails, the scheduler will retry, right?

yastij commented:
> I think this makes the most sense, otherwise we're just delaying the failure until it lands on a node. If the predicate fails, the scheduler will retry, right?

I agree here we'll just delay the failure, the predicate should fail and provide meaningful logs/reason.

bsalamat commented:
> I think this makes the most sense, otherwise we're just delaying the failure until it lands on a node. If the predicate fails, the scheduler will retry, right?

That's right. The scheduler will keep retrying (with some backoff time).

tallclair commented:
- Added failure modes for the scheduler & admission controller
- Renamed RuntimeController to RuntimeClass (admission controller)
- Marked KEP as implementable - I am happy with it's current state, and think we should commence coding.

tallclair commented:
Squashed commits.

derekwaynecarr commented on keps/sig-node/runtime-class-scheduling.md:
```diff
@@ -190,47 +186,64 @@ inculde:
+### RuntimeClass Admission Controller
 
-Merging tolerations is much simpler as we want to _union_ the RuntimeClass
+The RuntimeClass admission controller is a new in-tree admission plugin that
```
if the feature gate for runtime class is on, this will always be in the default set right?

tallclair replied:
Yes. Fixed.

derekwaynecarr commented:
/approve

we can clarify the admission requirement in a follow-on or during documentation.

yastij commented:
/lgtm
